polytracker/mapping.py

```diff
@@ -3,156 +3,191 @@
+                        if isinstance(nl, TDSourceNode):
+                            yield n.left
+                        else:
+                            stack.append(n.left)
+
+                    nr = self.tdfile.decode_node(n.right)
+                    if not nr.affects_control_flow:
+                        if isinstance(nr, TDSourceNode):
+                            yield n.right
+                        else:
+                            stack.append(n.right)
+
+                elif isinstance(n, TDRangeNode):
+                    for rl in range(n.first, n.last + 1):
+                        # NOTE: One could skip decoding here, but then we could end up with really long ranges
+                        # being added the labels that really does nothing except cause overhead...
+                        rn = self.tdfile.decode_node(rl)
+                        if rn.affects_control_flow:
+                            continue
+                        if isinstance(rn, TDSourceNode):
+                            yield rl
+                        else:
+                            stack.append(rl)
+
+        result: Dict[Path, List[CavityType]] = defaultdict(list)
+
+        for p, h in self.tdfile.fd_headers:
+            begin = h.prealloc_label_begin
+            end = h.prealloc_label_end
+            length = end - begin
+
+            if length < 1:
+                continue
+
+            marker = bytearray(length)
+            # Initially, mark all source taint that affects control flow
+            for i, label in enumerate(range(begin, end)):
+                if self.tdfile.decode_node(label).affects_control_flow:
+                    marker[i] = 1
+            # Now, iterate all source labels in the taint sink. As an optimization, if
+            # the taint affects_control_flow, move one. It already spilled into the source
+            # taint and was marked above
+            for sink in tqdm(list(self.tdfile.sinks)):
+                n = self.tdfile.decode_node(sink.label)
+                if n.affects_control_flow:
+                    continue
+                if isinstance(n, TDSourceNode):
+                    marker[n.offset] = 1
+                else:
+                    for source in source_labels_not_affecting_cf(sink.label):
+                        marker[source - begin] = 1
+
+            result[Path(p)] = self.marker_to_ranges(marker)
+
+        return result
 
 
 class MapInputsToOutputs(Command):
     name = "mapping"
     help = "generate a mapping of input byte offsets to output byte offsets"
 
     def __init_arguments__(self, parser):
-        parser.add_argument("POLYTRACKER_DB", type=str, help="the trace database")
+        parser.add_argument("POLYTRACKER_TF", type=str, help="the trace file")
 
     def run(self, args):
-        mapping = InputOutputMapping(PolyTrackerTrace.load(args.POLYTRACKER_DB)).mapping
-
-        print(mapping)
+        raise NotImplementedError()
```

Comment:
Any reason we couldn't keep this? I mean, if you wanted to get a mapping even though it is slower. Or does it just not make sense anymore?

Reply:
I plan on re-implementing it as my next step. Take out the DFS walk out of cavity detection and make it it's own method and use it in both `cavities` and `mapping`.